File: ios-samples/poslink/EposPosLinkExample/EposPosLinkExample/ViewModels/TransactionHistoryViewModel.swift
import Foundation

@Observable
final class TransactionHistoryViewModel {

    private(set) var refundingIds: Set<String> = []

    func refund(_ record: TransactionRecord) {
        guard let paymentId = record.gatewayPaymentId else { return }
        guard !refundingIds.contains(record.id) else { return }

        refundingIds.insert(record.id)
        TeyaService.shared.refundPayment(
            gatewayPaymentId: paymentId,
            amountMinor: Int32(record.amountMinor),
            currency: record.currency,
            onSettled: { [weak self] in
                DispatchQueue.main.async {
                    self?.refundingIds.remove(record.id)
                }
            }
        )
    }
}

File: android-samples/aio/EposAioExample/app/src/main/java/com/example/eposappexample/aio/ui/history/TransactionHistoryViewModel.kt
package com.example.eposappexample.aio.ui.history

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import com.example.eposappexample.aio.TeyaUtils
import com.example.eposappexample.aio.transactions.TransactionRecord
import com.example.eposappexample.aio.transactions.TransactionStore

class TransactionHistoryViewModel : ViewModel() {

    val transactions: List<TransactionRecord> get() = TransactionStore.transactions

    var refundingIds by mutableStateOf<Set<String>>(emptySet())
        private set

    fun refund(record: TransactionRecord) {
        val paymentId = record.gatewayPaymentId ?: return
        if (record.id in refundingIds) return

        refundingIds = refundingIds + record.id
        TeyaUtils.refundPayment(
            gatewayPaymentId = paymentId,
            amountMinor = record.amountMinor,
            currency = record.currency,
        ) {
            refundingIds = refundingIds - record.id
        }
    }
}
